Index: src/tcmalloc.cc
===================================================================
--- src/tcmalloc.cc	(revision 47)
+++ src/tcmalloc.cc	(working copy)
@@ -77,6 +77,7 @@
 #include <new>
 #include <stdio.h>
 #include <stddef.h>
+#include <time.h>
 #if defined HAVE_STDINT_H
 #include <stdint.h>
 #elif defined HAVE_INTTYPES_H
@@ -94,6 +95,10 @@
 #ifdef HAVE_UNISTD_H
 #include <unistd.h>
 #endif
+#ifdef HAVE_EXECINFO_H
+#include <execinfo.h>
+#endif
+
 #include <errno.h>
 #include <stdarg.h>
 #include "packed-cache-inl.h"
@@ -249,6 +254,22 @@
               "to return memory slower.  Reasonable rates are in the "
               "range [0,10]");
 
+DEFINE_uint64(tcmalloc_alloc_limit, 0,
+              "Limits the total alloc size. Will fail malloc/new if "
+              "allocation would cause to allocate more memory than"
+              "the limit");
+
+DEFINE_int32(tcmalloc_random_fail_freq, 0,
+              "rate at which we simulate memory failures");
+
+DEFINE_int32(tcmalloc_random_fail_skip_first, 0,
+              "when using tcmalloc_random_fail_freq, how many first allocs to skip");
+
+DEFINE_int32(tcmalloc_fill_mem, 0,
+              "if 1, will fill malloc()ed and free()d memory");
+
+static int until_next_alloc_fail;
+
 //-------------------------------------------------------------------
 // Mapping from size to size_class and vice versa
 //-------------------------------------------------------------------
@@ -276,6 +297,19 @@
 static const int add_amount[2] = { 7, 127 + (120 << 7) };
 static unsigned char class_array[377];
 
+int random_range(int lower, int upper)
+{
+    static bool initialized = false;
+    if (!initialized)
+    {
+        initialized = true;
+        srand((unsigned) time(NULL));
+    }
+    int range = (upper - lower);
+    int num = (int)(double(rand()) / (double(RAND_MAX) + 1.)) * range + lower;
+    return num;
+}
+
 // Compute index of the class_array[] entry for a given size
 static inline int ClassIndex(int s) {
   ASSERT(0 <= s);
@@ -1430,6 +1464,7 @@
   pthread_t     tid_;                   // Which thread owns it
   FreeList      list_[kNumClasses];     // Array indexed by size-class
   bool          in_setspecific_;        // In call to pthread_setspecific?
+  bool		exempt_;		// (Temporarily) exempt from induced memory allocation failures
 
   // Allocate a new heap. REQUIRES: pageheap_lock is held.
   static inline TCMalloc_ThreadCache* NewHeap(pthread_t tid);
@@ -1470,6 +1505,10 @@
   // Pick next sampling point
   void PickNextSample(size_t k);
 
+  // Exempting from induced memory allocation failures
+  bool isExempt(void);
+  void setExempt(bool b);
+
   static void                  InitModule();
   static void                  InitTSD();
   static TCMalloc_ThreadCache* GetThreadHeap();
@@ -1924,6 +1963,7 @@
   prev_ = NULL;
   tid_  = tid;
   in_setspecific_ = false;
+  exempt_ = false;
   for (size_t cl = 0; cl < kNumClasses; ++cl) {
     list_[cl].Init();
   }
@@ -2282,6 +2322,16 @@
   }
 }
 
+bool TCMalloc_ThreadCache::isExempt(void)
+{
+	return exempt_;
+}
+
+void TCMalloc_ThreadCache::setExempt(bool b)
+{
+	exempt_ = b;
+}
+
 // Extract interesting stats
 struct TCMallocStats {
   uint64_t system_bytes;        // Bytes alloced from system
@@ -2327,6 +2377,21 @@
   }
 }
 
+// Returned a number of bytes currently allocated by the app
+// TODO: it's slow. I was thinking about keeping a running count and updating
+// it in do_malloc() and do_free(), but couldn't figure out how to get the
+// size of allocation from the pointer in do_free()
+static uint64_t GetAllocatedBytes() {
+  TCMallocStats stats;
+  ExtractStats(&stats, NULL);
+  const uint64_t bytes_in_use = stats.system_bytes
+                                - stats.pageheap_bytes
+                                - stats.central_bytes
+                                - stats.transfer_bytes
+                                - stats.thread_bytes;
+  return bytes_in_use;
+}
+
 // WRITE stats to "out"
 static void DumpStats(TCMalloc_Printer* out, int level) {
   TCMallocStats stats;
@@ -2600,6 +2665,46 @@
     TCMalloc_ThreadCache::InitTSD();
     free(malloc(1));
     MallocExtension::Register(new TCMallocImplementation);
+
+    const char* env = getenv("MALLOCLIMIT");
+    if (env != NULL) {
+      int limit = atoi(env);
+      if (limit > 0) {
+        FLAGS_tcmalloc_alloc_limit = limit;
+        printf("Setting malloc limit to %d\n", limit);
+      }
+    }
+
+    env = getenv("MALLOCFAILRATE");
+    if (env != NULL) {
+      int limit = atoi(env);
+      if (limit > 0) {
+        FLAGS_tcmalloc_random_fail_freq = limit;
+        printf("Setting malloc fail rate %d\n", FLAGS_tcmalloc_random_fail_freq);
+        until_next_alloc_fail = random_range(1, FLAGS_tcmalloc_random_fail_freq);
+        printf("Will fail next %d allocation.\n", until_next_alloc_fail);
+      }
+    }
+
+    env = getenv("MALLOCFAILSKIPFIRST");
+    if (env != NULL) {
+      int limit = atoi(env);
+      if (limit > 0) {
+        FLAGS_tcmalloc_random_fail_skip_first = limit;
+        printf("Setting malloc fail skip first to %d\n", limit);
+      }
+    }
+
+    env = getenv("MALLOCFILLMEM");
+    if (env != NULL) {
+      int fill = atoi(env);
+      FLAGS_tcmalloc_fill_mem = fill;
+      if (fill)
+        printf("Setting malloc to fill malloc()ed and free()d memory\n");
+      else
+        printf("Setting malloc to NOT fill malloc()ed and free()d memory\n");           
+    }
+  
   }
 
   ~TCMallocGuard() {
@@ -2678,11 +2783,59 @@
   return span == NULL ? NULL : SpanToMallocResult(span);
 }
 
+static const int kMaxStackTrace = 32;
+
+static void dump_callstack() {
+    void* key[kMaxStackTrace];
+    int depth = MallocHook::GetCallerStackTrace(key, kMaxStackTrace, 2);
+#ifdef HAVE_EXECINFO_H
+    char ** syms = backtrace_symbols(key, depth);
+    for (int i = 0; key && (i < depth); i++) {
+      printf("%p, %s\n", key[i], syms[i]);
+    }
+    free(syms);
+#else
+    for (int i = 0; key && (i < depth); i++) {
+      printf("%p\n", key[i]);
+    }
+#endif
+}
+
+void tcmalloc_set_exempt(bool b)
+{
+  TCMalloc_ThreadCache* heap = TCMalloc_ThreadCache::GetCache();
+  heap->setExempt(b);
+}
+
 static inline void* do_malloc(size_t size) {
   void* ret = NULL;
-
   // The following call forces module initialization
   TCMalloc_ThreadCache* heap = TCMalloc_ThreadCache::GetCache();
+  if ((FLAGS_tcmalloc_alloc_limit > 0) && !heap->isExempt() && (GetAllocatedBytes() + size) > FLAGS_tcmalloc_alloc_limit) {
+    heap->setExempt(true);
+    printf("\nFailing allocation of size %d because reached allocation limit. Callstack:\n", (int)size);
+    dump_callstack();
+    heap->setExempt(false);
+    errno = ENOMEM;
+    return NULL;
+  }
+
+  if (FLAGS_tcmalloc_random_fail_freq != 0 && !heap->isExempt()) {
+    if (0 == FLAGS_tcmalloc_random_fail_skip_first) {
+      if (0 == until_next_alloc_fail--) {
+        heap->setExempt(true);
+        until_next_alloc_fail = random_range(1, FLAGS_tcmalloc_random_fail_freq);
+        printf("\nRandomly failing allocation of size %d on thread %d. Will fail next %d allocation. Callstack:\n", (int)size, (int) pthread_self(), until_next_alloc_fail);
+        dump_callstack();
+        heap->setExempt(false);
+        errno = ENOMEM;
+        return NULL;
+      }
+    } else {
+      --FLAGS_tcmalloc_random_fail_skip_first;
+    }
+  }
+
   if ((FLAGS_tcmalloc_sample_parameter > 0) && heap->SampleAllocation(size)) {
     Span* span = DoSampledAllocation(size);
     if (span != NULL) {
@@ -2695,7 +2848,12 @@
   } else {
     ret = do_malloc_pages(pages(size));
   }
+
   if (ret == NULL) errno = ENOMEM;
+
+  if (ret && FLAGS_tcmalloc_fill_mem) {
+    memset(ret, 0xCD, size);
+  }
   return ret;
 }
 
@@ -2715,6 +2873,12 @@
     ASSERT(!pageheap->GetDescriptor(p)->sample);
     TCMalloc_ThreadCache* heap = TCMalloc_ThreadCache::GetCacheIfPresent();
     if (heap != NULL) {
+      // important to call *before* Deallocate, because first 4 bytes of ptr
+      // are re-used for linked list pointer
+      if (FLAGS_tcmalloc_fill_mem) {
+        size_t size = ByteSizeForClass(cl);
+        memset(ptr, 0xdd, size);
+      }
       heap->Deallocate(ptr, cl);
     } else {
       // Delete directly into central cache
